File: apps/ios/App/Screens/SharedLogbookScreen.swift
import SwiftUI
import TrackEvolutionKit

/// Someone's public logbook, read-only — where a `trackevolution.app/share/<slug>`
/// Universal Link lands.
///
/// This screen exists because that link pattern is the only one the association file
/// advertises (`src/routes/wellKnown.ts`), so without it tapping a shared link opens
/// the app showing *your* logbook instead of theirs, which is worse than not handling
/// the link at all.
///
/// It reads `GET /api/share/:slug`, which is **unauthenticated** and already strips
/// notes, email, per-lap data and every garage/setup linkage server-side. `ShareData`
/// is a separate model from `Event`/`Track` for the same reason: a private field
/// leaking into the public payload is a decode failure here, not a silent nil.
struct SharedLogbookScreen: View {
    let slug: String

    @Environment(AuthController.self) private var auth
    @State private var state: LoadState = .loading
    @State private var data: ShareData?

    var body: some View {
        TELoadable(state: state, retry: load) {
            if let data {
                content(data)
            }
        }
        .navigationTitle(data?.name ?? slug)
        .navigationBarTitleDisplayMode(.inline)
        .task {
            if data == nil { await load() }
        }
    }

    private func content(_ data: ShareData) -> some View {
        TEPage {
            VStack(alignment: .leading, spacing: 4) {
                Text(data.name ?? "Shared logbook")
                    .teStyle(.h1)
                    .foregroundStyle(Color(.textStrong))
                Text("A read-only track history — notes and per-lap data stay private.")
                    .teStyle(.xs)
                    .foregroundStyle(Color(.textMuted))
            }

            TEStatRow(tiles: [
                TEStatTile(label: "Events", value: "\(data.totals.events)"),
                TEStatTile(label: "Track days", value: "\(data.totals.trackDays)"),
                TEStatTile(label: "Tracks", value: "\(data.tracks.count)")
            ])

            TESectionHeader("Tracks")
            if data.tracks.isEmpty {
                TEEmpty("No tracks on this page yet.")
            } else {
                ForEach(data.tracks) { track in
                    TECard(padding: 14) {
                        VStack(alignment: .leading, spacing: 6) {
                            HStack(alignment: .firstTextBaseline) {
                                Text(track.name)
                                    .teStyle(.h3)
                                    .foregroundStyle(Color(.textStrong))
                                Spacer()
                                TETime(ms: track.bestMs)
                            }
                            TEMeta([
                                fmtCount(track.eventCount, "event"),
                                fmtCount(track.trackDays, "day"),
                                EventDates.fmtDate(track.lastDate)
                            ])
                            if track.series.count >= 2 {
                                ProgressChart(
                                    points: track.series.enumerated().map { index, point in
                                        .init(
                                            x: Double(index),
                                            label: EventDates.fmtDate(point.date),
                                            ms: point.bestMs
                                        )
                                    },
                                    style: .sparkline
                                )
                                // Laid out full-width under the text rather than beside
                                // it, so nothing else fixes its height.
                                .frame(height: 44)
                            }
                        }
                    }
                }
            }

            TESectionHeader("Events")
            ForEach(data.events) { event in
                TECard(padding: 14) {
                    HStack(alignment: .firstTextBaseline) {
                        VStack(alignment: .leading, spacing: 4) {
                            Text(event.trackName)
                                .teStyle(.bodyStrong)
                                .foregroundStyle(Color(.textStrong))
                            TEMeta([
                                EventDates.fmtDate(event.startDate),
                                event.club,
                                event.runGroup,
                                event.conditions?.label
                            ])
                        }
                        Spacer(minLength: 8)
                        VStack(alignment: .trailing, spacing: 2) {
                            TETime(ms: event.bestMs)
                            Text(LapTime.fmtConsistency(event.consistency))
                                .teStyle(.xxs)
                                .foregroundStyle(Color(.textFaint))
                        }
                    }
                }
            }
        }
        .refreshable { await load() }
    }

    private func load() async {
        do {
            data = try await auth.api.sharedLogbook(slug: slug)
            state = .ready
        } catch let error as APIError where error.status == 404 {
            state = .failed("There's no shared logbook at /share/\(slug) — the link may have been disabled.")
        } catch let error as APIError {
            state = .failed(error.message)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/screens/SharedLogbookScreen.kt
package app.trackevolution.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import app.trackevolution.core.EventDates
import app.trackevolution.core.LapTime
import app.trackevolution.core.api.ApiClient
import app.trackevolution.core.api.ApiException
import app.trackevolution.core.model.ShareData
import app.trackevolution.ui.LoadState
import app.trackevolution.ui.TEEmpty
import app.trackevolution.ui.fmtCount
import app.trackevolution.ui.TELoadable
import app.trackevolution.ui.TEMeta
import app.trackevolution.ui.TESectionHeader
import app.trackevolution.ui.TEStatRow
import app.trackevolution.ui.charts.ProgressChart
import app.trackevolution.ui.charts.ProgressChartStyle
import app.trackevolution.ui.charts.ProgressPoint
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/**
 * Somebody else's logbook, read-only (`GET /api/share/:slug`).
 *
 * This is where an App Link lands — `/share/<slug>` is the only pattern
 * `src/routes/wellKnown.ts` advertises — so it has to exist even though it is
 * not one of NS-26's five screens: without it a shared link opens a blank app,
 * which is precisely the dead end the spec rules out.
 *
 * **Nothing here is tappable through to a private screen.** The payload is a
 * different shape from the authed one on purpose (`SharedEvent`/`SharedTrack`
 * carry no notes, no laps, no garage linkage), and a row that navigated into the
 * signed-in logbook would be showing the wrong person's data.
 */
class SharedLogbookModel(
    private val scope: CoroutineScope,
    private val api: ApiClient,
    val slug: String,
) {
    var state by mutableStateOf<LoadState>(LoadState.Loading)
        private set

    var data by mutableStateOf<ShareData?>(null)
        private set

    fun load() {
        scope.launch {
            try {
                data = api.sharedLogbook(slug)
                state = LoadState.Ready
            } catch (e: ApiException) {
                state = LoadState.Failed(e.message ?: "That shared logbook isn't available.")
            }
        }
    }
}

@Composable
fun SharedLogbookScreen(model: SharedLogbookModel, modifier: Modifier = Modifier) {
    val colors = TrackTheme.colors
    val listState = rememberLazyListState()

    LaunchedEffect(Unit) { if (model.state == LoadState.Loading) model.load() }

    TELoadable(state = model.state, onRetry = model::load, modifier = modifier) {
        val data = model.data ?: return@TELoadable

        LazyColumn(
            state = listState,
            modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
            contentPadding = PaddingValues(vertical = 12.dp),
        ) {
            item("head") {
                Column {
                    Text(data.name ?: "Shared logbook", style = TrackTheme.typography.h1, color = colors.textStrong)
                    Text("A read-only track history. Notes and per-lap data stay private.", style = TrackTheme.typography.sm, color = colors.textMuted)
                }
            }

            item("totals") {
                TEStatRow(
                    listOf(
                        "Events" to data.totals.events.toString(),
                        "Track days" to data.totals.trackDays.toString(),
                        "Tracks" to data.tracks.size.toString(),
                    ),
                )
            }

            item("tracks-header") { TESectionHeader("Tracks") }
            if (data.tracks.isEmpty()) {
                item("tracks-empty") { TEEmpty("Nothing logged here yet.") }
            } else {
                items(data.tracks, key = { "tr-${it.id}" }) { track ->
                    TrackCard(Modifier.fillMaxWidth()) {
                        Row(
                            modifier = Modifier.fillMaxWidth().padding(bottom = 4.dp),
                            horizontalArrangement = Arrangement.SpaceBetween,
                        ) {
                            Column(Modifier.weight(1f)) {
                                Text(track.name, style = TrackTheme.typography.bodyStrong, color = colors.textStrong)
                                TEMeta(
                                    listOf(
                                        fmtCount(track.eventCount, "event"),
                                        fmtCount(track.trackDays, "day"),
                                        EventDates.fmtDate(track.lastDate),
                                    ),
                                )
                            }
                            Text(LapTime.fmtMs(track.bestMs), style = TrackTheme.typography.lapTime, color = colors.textStrong)
                        }
                        if (track.series.size >= 2) {
                            ProgressChart(
                                points = track.series.mapIndexed { i, point ->
                                    ProgressPoint(x = i.toDouble(), label = "", ms = point.bestMs)
                                },
                                style = ProgressChartStyle.Sparkline,
                            )
                        }
                    }
                }
            }

            item("events-header") { TESectionHeader("Events") }
            items(data.events, key = { "ev-${it.id}" }) { event ->
                TrackCard(Modifier.fillMaxWidth()) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                    ) {
                        Column(Modifier.weight(1f)) {
                            Text(event.trackName, style = TrackTheme.typography.bodyStrong, color = colors.textStrong)
                            TEMeta(
                                listOf(
                                    EventDates.fmtDate(event.startDate),
                                    event.club,
                                    event.runGroup,
                                    event.conditions?.rawValue,
                                ),
                            )
                        }
                        Text(LapTime.fmtMs(event.bestMs), style = TrackTheme.typography.lapTime, color = colors.textStrong)
                    }
                }
            }
        }
    }
}
